misc/python/materialize/data_ingest/workload.py

```diff
@@ -0,0 +1,152 @@
+
+import pg8000
+
+from materialize.data_ingest.data_type import (
+    DoubleType,
+    FloatType,
+    IntType,
+    LongType,
+    StringType,
+)
+from materialize.data_ingest.definition import (
+    Definition,
+    Delete,
+    Insert,
+    Keyspace,
+    Records,
+    RecordSize,
+    Upsert,
+)
+from materialize.data_ingest.executor import PgExecutor, PrintExecutor
+from materialize.data_ingest.field import Field
+from materialize.data_ingest.transaction import Transaction
+
+
+class Workload:
+    cycle: List[Definition]
+    num_cycles: int
+
+    def generate(self, fields: List[Field]) -> List[Transaction]:
+        transactions = []
+        for i in range(self.num_cycles):
+            cycle = deepcopy(self.cycle)
+            for definition in cycle:
+                transactions.extend(definition.generate(fields))
+        return transactions
+
+
+class SingleSensorUpdating(Workload):
+    def __init__(self, scale_factor: int) -> None:
+        self.num_cycles = 10 * scale_factor
+        self.cycle: List[Definition] = [
+            Upsert(
+                keyspace=Keyspace.SINGLE_VALUE,
+                count=Records.ONE,
+                record_size=RecordSize.SMALL,
+            )
+        ]
+
+
+class DeleteDataAtEndOfDay(Workload):
+    def __init__(self, scale_factor: int) -> None:
+        self.num_cycles = scale_factor
+        insert = Insert(
+            count=Records.ONE,
+            record_size=RecordSize.SMALL,
+        )
+        self.cycle: List[Definition] = [
+            insert,
+            Delete(
+                number_of_records=Records.ALL,
+                record_size=RecordSize.SMALL,
+                num=insert.max_key(),
+            ),
+        ]
+
+
+class ProgressivelyEnrichRecords(Workload):
+    def __init__(self, scale_factor: int) -> None:
+        self.num_cycles = scale_factor
+        self.cycle: List[Definition] = [
+            # TODO: Implement
+        ]
+
+
+# TODO: Disruptions in workloads
+
+
+def execute_workload(
+    executor_classes: Any,
+    workload: Workload,
+    conn: pg8000.Connection,
+    num: int,
+    ports: Dict[str, int],
+    verbose: bool,
+) -> None:
+    fields = []
+
+    for i in range(random.randint(1, 10)):
+        typ = random.choice((StringType, IntType, LongType, FloatType, DoubleType))
+        fields.append(Field(f"key{i}", typ, True))
+    for i in range(random.randint(0, 20)):
+        typ = random.choice((StringType, IntType, LongType, FloatType, DoubleType))
+        fields.append(Field(f"value{i}", typ, False))
+    print(f"With fields: {fields}")
+
+    transactions = workload.generate(fields)
+    # print(transactions)
+
+    pg_executor = PgExecutor(num, ports, fields)
+    executors = [
+        executor_class(num, conn, ports, fields) for executor_class in executor_classes
+    ]
+    run_executors = [pg_executor] + executors
+    if verbose:
+        run_executors = [PrintExecutor()] + executors
+
+    for executor in run_executors:
```

Comment:
I think this loop nesting needs to be backwards, for every transaction, send it to both executors.